src/Control/FunFlow/External/Executor.hs

```diff
@@ -0,0 +1,95 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+-- | Executor for external tasks.
+module Control.FunFlow.External.Executor where
+
+import           Control.Exception                    (IOException, try)
+import           Control.FunFlow.ContentStore
+import           Control.FunFlow.External
+import           Control.FunFlow.External.Coordinator
+import           Control.Lens
+import           Control.Monad                        (forever)
+import qualified Data.Text                            as T
+import           Network.HostName
+import           System.Clock
+import           System.Exit                          (ExitCode (..))
+import           System.FilePath                      ((</>))
+import           System.IO                            (IOMode (..), openFile)
+import           System.Process
+
+data ExecutionResult =
+    -- | The result already exists in the store and there is no need
+    --   to execute. This is also returned if the job is already running
+    --   elsewhere.
+    Cached
+    -- | Execution completed successfully after a certain amount of time.
+  | Success TimeSpec
+    -- | Execution failed with the following exit code.
+    --   TODO where should logs go?
+  | Failure TimeSpec Int
+
+-- | Execute an individual task.
+execute :: ContentStore -> TaskDescription -> IO ExecutionResult
+execute store td = do
+  instruction <- constructIfMissing store (td ^. tdOutput)
+  case instruction of
+    Wait -> return Cached
+    Consume _ -> return Cached
+    Construct fp -> let
+        defaultProc = proc (T.unpack $ td ^. tdTask . etCommand)
+                       (T.unpack <$> td ^. tdTask . etParams)
+        procSpec out = defaultProc {
+            cwd = Just fp
+          , close_fds = True
+            -- Error output should be displayed on our stderr stream
+          , std_err = Inherit
+          , std_out = out
+          }
+      in do
+        out <-
+          if (td ^. tdTask . etWriteToStdOut)
+          then UseHandle <$> openFile (fp </> "out") WriteMode
+          else return Inherit
+
+        start <- getTime Monotonic
+        mp <- try $ createProcess $ procSpec out
+        case mp of
+          Left (_ex :: IOException) -> do
+            removeFailed store (td ^. tdOutput)
+            return $ Failure (diffTimeSpec start start) 2
+          Right (_, _, _, ph) -> do
+            exitCode <- waitForProcess ph
+            end <- getTime Monotonic
+            case exitCode of
+              ExitSuccess   -> do
```

Comment:
This is the only path that leads to a store entry being marked as complete, right?
But how to handle the case where ffexecutord is restarted while external processes are running?
It looks like currently `execute` would mark these external tasks as `Cached` and execute loop would then mark the status as `Completed`.